File: src/potential/polyprism.c
/* **************************************************************************

   Calculate the potential fields and derivatives of the 3D prism with polygonal
   crossection. Uses forumla of Plouff (1976)

   Author: Vanderlei Coelho de Olivera Junior

   ************************************************************************** */

#include <math.h>


const double GRAV = 0.00000000006673; /* Gravitational constant in SI */


/* The following functions calculate the gravitational potential and its first
and second derivatives caused by a prism with polygonal crossection
using the formula given in Plouff (1976).

The coordinate system of the input parameters is assumed to be
    x->north, y->east; z->down.

Input values in SI units and returns values in:
* gx, gy, gz = mGal
* gxx, gxy, gxz, gyy, etc. = Eotvos

Parameters:
    * double dens: density of the prism
    * double z1, z2: top and bottom of the prism
    * double *x, *y: x and y coordinates of the vertices of the prism
    * int nvertices: number of vertices
    * double *xp, *xp, *zp: coordinates of the computation points
    * int n: number of computation points
    * double *res: vector used to return the calculated effect on the n points
Returns:
    * int: 0
*/
int polyprism_gz(double dens, double z1, double z2, double *x, double *y,
                 int nvertices, double *xp, double *yp, double *zp, int n,
				 double *res)
{
	register int i, k;
	double gzaux, aux;
	double Xk1, Xk2, Yk1, Yk2, Z1, Z1_quadrado, Z2, Z2_quadrado;
	double auxk1, auxk2, aux1k1, aux1k2, aux2k1, aux2k2;
	double Ak1, Ak2, Bk1, Bk2, Ck1, Ck2, Dk1, Dk2, E1k1, E1k2, E2k1, E2k2;
	double Qk1, Qk2, R1k1, R1k2, R2k1, R2k2, p, p_quadrado;
    double *p2x, *p2y; /* pointers to x and y */

	for (i = 0; i < n; i++)
    {
		*res = 0.0;
        /* 100000 transforms SI to mGal */
        aux = GRAV*dens*100000.0;
        gzaux = 0.0;
        Z1 = z1 - *zp;
        Z2 = z2 - *zp;
        Z1_quadrado = pow(Z1, 2);
        Z2_quadrado = pow(Z2, 2);
        /* Loop over vertices */
        p2x = x;
        p2y = y;
        for (k = 0; k < (nvertices-1); k++)
        {
            Xk1 = *p2x - *xp;
            Xk2 = *(p2x+1) - *xp;
            Yk1 = *p2y - *yp;
            Yk2 = *(p2y+1) - *yp;

            p = (Xk1*Yk2) - (Xk2*Yk1);
            p_quadrado = pow(p, 2);
            Qk1 = ((Yk2 - Yk1)*Yk1) + ((Xk2 - Xk1)*Xk1);
            Qk2 = ((Yk2 - Yk1)*Yk2) + ((Xk2 - Xk1)*Xk2);
            Ak1 = pow(Xk1, 2) + pow(Yk1, 2);
            Ak2 = pow(Xk2, 2) + pow(Yk2, 2);
            R1k1 = Ak1 + Z1_quadrado;
            R1k1 = pow(R1k1, 0.5);
            R1k2 = Ak2 + Z1_quadrado;
            R1k2 = pow(R1k2, 0.5);
            R2k1 = Ak1 + Z2_quadrado;
            R2k1 = pow(R2k1, 0.5);
            R2k2 = Ak2 + Z2_quadrado;
            R2k2 = pow(R2k2, 0.5);
            Ak1 = pow(Ak1, 0.5);
            Ak2 = pow(Ak2, 0.5);
            Bk1 = pow(Qk1, 2) + p_quadrado;
            Bk1 = pow(Bk1, 0.5);
            Bk2 = pow(Qk2, 2) + p_quadrado;
            Bk2 = pow(Bk2, 0.5);
            Ck1 = Qk1*Ak1;
            Ck2 = Qk2*Ak2;
            #define Divide_macro(a, b) ((a)/((b) + (1E-10)))
            Dk1 = Divide_macro(p, 2.0);
            Dk2 = Dk1;
            Dk1 *= Divide_macro(Ak1, Bk1);
            Dk2 *= Divide_macro(Ak2, Bk2);
            E1k1 = R1k1*Bk1;
            E1k2 = R1k2*Bk2;
            E2k1 = R2k1*Bk1;
            E2k2 = R2k2*Bk2;
            auxk1 = Divide_macro(Qk1, p);
            auxk2 = Divide_macro(Qk2, p);
            aux1k1 = Divide_macro(Z1, R1k1);
            aux1k2 = Divide_macro(Z1, R1k2);
            aux2k1 = Divide_macro(Z2, R2k1);
            aux2k2 = Divide_macro(Z2, R2k2);
            #undef Divide_macro
            gzaux += (Z2 - Z1)*(atan(auxk2) - atan(auxk1));
            gzaux += Z2*(atan(aux2k1*auxk1) - atan(aux2k2*auxk2));
            gzaux += Z1*(atan(aux1k2*auxk2) - atan(aux1k1*auxk1));
            #define Divide_macro2(a, b) (((a) + (1E-10))/((b) + (1E-10)))
            gzaux += Dk1*(log(Divide_macro2((E1k1 - Ck1), (E1k1 + Ck1))) -
                          log(Divide_macro2((E2k1 - Ck1), (E2k1 + Ck1))));
            gzaux += Dk2*(log(Divide_macro2((E2k2 - Ck2), (E2k2 + Ck2))) -
                          log(Divide_macro2((E1k2 - Ck2), (E1k2 + Ck2))));
            #undef Divide_macro2
            p2x++;
            p2y++;
        }
        /* Calculate the last edge */
        Xk1 = *p2x - *xp;
        Xk2 = *x - *xp;
        Yk1 = *p2y - *yp;
        Yk2 = *y - *yp;

        p = (Xk1*Yk2) - (Xk2*Yk1);
        p_quadrado = pow(p, 2);
        Qk1 = ((Yk2 - Yk1)*Yk1) + ((Xk2 - Xk1)*Xk1);
        Qk2 = ((Yk2 - Yk1)*Yk2) + ((Xk2 - Xk1)*Xk2);
        Ak1 = pow(Xk1, 2) + pow(Yk1, 2);
        Ak2 = pow(Xk2, 2) + pow(Yk2, 2);
        R1k1 = Ak1 + Z1_quadrado;
        R1k1 = pow(R1k1, 0.5);
        R1k2 = Ak2 + Z1_quadrado;
        R1k2 = pow(R1k2, 0.5);
        R2k1 = Ak1 + Z2_quadrado;
        R2k1 = pow(R2k1, 0.5);
        R2k2 = Ak2 + Z2_quadrado;
        R2k2 = pow(R2k2, 0.5);
        Ak1 = pow(Ak1, 0.5);
        Ak2 = pow(Ak2, 0.5);
        Bk1 = pow(Qk1, 2) + p_quadrado;
        Bk1 = pow(Bk1, 0.5);
        Bk2 = pow(Qk2, 2) + p_quadrado;
        Bk2 = pow(Bk2, 0.5);
        Ck1 = Qk1*Ak1;
        Ck2 = Qk2*Ak2;
        #define Divide_macro(a, b) ((a)/((b) + (1E-10)))
        Dk1 = Divide_macro(p, 2.0);
        Dk2 = Dk1;
        Dk1 *= Divide_macro(Ak1, Bk1);
        Dk2 *= Divide_macro(Ak2, Bk2);
        E1k1 = R1k1*Bk1;
        E1k2 = R1k2*Bk2;
        E2k1 = R2k1*Bk1;
        E2k2 = R2k2*Bk2;
        auxk1 = Divide_macro(Qk1, p);
        auxk2 = Divide_macro(Qk2, p);
        aux1k1 = Divide_macro(Z1, R1k1);
        aux1k2 = Divide_macro(Z1, R1k2);
        aux2k1 = Divide_macro(Z2, R2k1);
        aux2k2 = Divide_macro(Z2, R2k2);
        #undef Divide_macro
        gzaux += (Z2 - Z1)*(atan(auxk2) - atan(auxk1));
        gzaux += Z2*(atan(aux2k1*auxk1) - atan(aux2k2*auxk2));
        gzaux += Z1*(atan(aux1k2*auxk2) - atan(aux1k1*auxk1));
        #define Divide_macro2(a, b) (((a) + (1E-10))/((b) + (1E-10)))
        gzaux += Dk1*(log(Divide_macro2((E1k1 - Ck1), (E1k1 + Ck1))) -
                      log(Divide_macro2((E2k1 - Ck1), (E2k1 + Ck1))));
        gzaux += Dk2*(log(Divide_macro2((E2k2 - Ck2), (E2k2 + Ck2))) -
                      log(Divide_macro2((E1k2 - Ck2), (E1k2 + Ck2))));
        #undef Divide_macro
        gzaux *= aux;
        *res += gzaux;
        res++;
        xp++;
        yp++;
        zp++;
	}
	return 0;
}

File: fatiando/potential/_polyprism.py
"""
.. topic:: Python + Numpy implementation.

    Module :mod:`fatiando.potential.polyprism` loads all functions from
    ``fatiando.potential._polyprism``, which contain the Python + Numpy
    implementation. The slightly faster Cython module
    ``fatiando.potentia._cpolyprism`` will be out soon. If it is available,
    then will substitude the Python + Numpy functions with its functions. All
    input and output are the same but there is a slight speed increase.
    
"""
import numpy
from numpy import arctan2, log, sqrt, arctan

from fatiando import logger
from fatiando.constants import SI2MGAL, SI2EOTVOS, G


def gz(xp, yp, zp, prisms):
    """
    Calculates the :math:`g_z` gravity acceleration component.

    .. note:: The coordinate system of the input parameters is to be x -> North,
        y -> East and z -> **DOWN**.

    .. note:: All input values in **SI** units(!) and output in **mGal**!

    Parameters:
    
    * xp, yp, zp : arrays
        The x, y, and z coordinates of the computation points.
    * prisms : listSI2MGAL
        List of :class:`fatiando.mesher.ddd.PolygonalPrism` objects.

    Returns:
    
    * res : array
        The effect calculated on the computation points.

    """
    if xp.shape != yp.shape != zp.shape:
        raise ValueError("Input arrays xp, yp, and zp must have same shape!")
    dummy = 10**(-10)
    res = numpy.zeros(len(xp), dtype='f')
    for prism in prisms:
        if prism is None or 'density' not in prism.props:
            continue
        density = prism.props['density']
        nverts = prism.nverts
        x, y = prism.x, prism.y
        z1, z2 = prism.z1, prism.z2
        # Calculate the effect of the prism
        Z1 = z1 - zp
        Z2 = z2 - zp
        Z1_sqr = Z1**2
        Z2_sqr = Z2**2
        kernel = numpy.zeros_like(res)
        for k in range(nverts):
            Xk1 = x[k] - xp
            Yk1 = y[k] - yp
            Xk2 = x[(k + 1)%nverts] - xp
            Yk2 = y[(k + 1)%nverts] - yp
            p = Xk1*Yk2 - Xk2*Yk1
            p_sqr = p**2
            Qk1 = (Yk2 - Yk1)*Yk1 + (Xk2 - Xk1)*Xk1
            Qk2 = (Yk2 - Yk1)*Yk2 + (Xk2 - Xk1)*Xk2
            Ak1 = Xk1**2 + Yk1**2
            Ak2 = Xk2**2 + Yk2**2
            R1k1 = sqrt(Ak1 + Z1_sqr)
            R1k2 = sqrt(Ak2 + Z1_sqr)
            R2k1 = sqrt(Ak1 + Z2_sqr)
            R2k2 = sqrt(Ak2 + Z2_sqr)
            Ak1 = sqrt(Ak1)
            Ak2 = sqrt(Ak2)
            Bk1 = sqrt(Qk1**2 + p_sqr)
            Bk2 = sqrt(Qk2**2 + p_sqr)
            E1k1 = R1k1*Bk1
            E1k2 = R1k2*Bk2
            E2k1 = R2k1*Bk1
            E2k2 = R2k2*Bk2
            kernel += (Z2 - Z1)*(arctan2(Qk2, p) - arctan2(Qk1, p))
            kernel += Z2*(arctan2(Z2*Qk1, R2k1*p) - arctan2(Z2*Qk2, R2k2*p))
            kernel += Z1*(arctan2(Z1*Qk2, R1k2*p) - arctan2(Z1*Qk1, R1k1*p))
            Ck1 = Qk1*Ak1
            Ck2 = Qk2*Ak2
            # dummy helps prevent zero division errors
            kernel += 0.5*p*(Ak1/(Bk1 + dummy))*(
                log((E1k1 - Ck1)/(E1k1 + Ck1 + dummy) + dummy) -
                log((E2k1 - Ck1)/(E2k1 + Ck1 + dummy) + dummy))
            kernel += 0.5*p*(Ak2/(Bk2 + dummy))*(
                log((E2k2 - Ck2)/(E2k2 + Ck2 + dummy) + dummy) -
                log((E1k2 - Ck2)/(E1k2 + Ck2 + dummy) + dummy))
        res = res + kernel*density
    res *= G*SI2MGAL
    return res 

def gxx(xp, yp, zp, prisms):
    """
    Calculates the :math:`g_xx` gravity gradient tensor component.

    .. note:: The coordinate system of the input parameters is to be x -> North,
        y -> East and z -> **DOWN**.

    .. note:: All input values in **SI** units(!) and output in **mGal**!

    Parameters:
    
    * xp, yp, zp : arrays
        The x, y, and z coordinates of the computation points.
    * prisms : list
        List of :class:`fatiando.mesher.ddd.PolygonalPrism` objects.

    Returns:
    
    * res : array
        The effect calculated on the computation points.

    """
    if xp.shape != yp.shape != zp.shape:
        raise ValueError("Input arrays xp, yp, and zp must have same shape!")
    dummy = 10**(-10)
    res = numpy.zeros(len(xp), dtype='f')
    for prism in prisms:
        if prism is None or 'density' not in prism.props:
            continue
        density = prism.props['density']
        nverts = prism.nverts
        x, y = prism.x, prism.y
        z1, z2 = prism.z1, prism.z2
        # Calculate the effect of the prism
        Z1 = z1 - zp
        Z2 = z2 - zp
        for k in range(nverts):
            res += density*_integral_v1(x[k] - xp, x[(k + 1)%nverts] - xp,
                y[k] - yp, y[(k + 1)%nverts] - yp, Z1, Z2)
    res *= G*SI2EOTVOS
    return res
                    
def _integral_v1(x1, x2, y1, y2, z1, z2):
    """
    Calculates the first element of the V matrix (gxx components)
    """    
    dummy = 10.**(-10) # Used to avoid singularities
    dx = x2 - x1
    dy = y2 - y1
    dr = sqrt(dx**2 + dy**2)
    z1_sqr = z1**2
    z2_sqr = z2**2
    n = dx/(dy + dummy)
    g = x1 - y1*n
    p = (x1*y2 - x2*y1)/(dr + dummy)    
    
    r2_sqr = x2**2 + y2**2
    R21 = sqrt(r2_sqr + z1_sqr)
    R22 = sqrt(r2_sqr + z2_sqr)
    d2 = (dx*x2 + dy*y2)/(dr + dummy)
    aux = arctan2(z2*d2, p*R22) - arctan2(z1*d2, p*R21)
    res = g*y2*aux/(p*d2 + dummy) + n*p*aux/(d2 + dummy)
    
    r1_sqr = x1**2 + y1**2
    R11 = sqrt(r1_sqr + z1_sqr)
    R12 = sqrt(r1_sqr + z2_sqr)
    d1 = dx*x1 + dy*y1/(dr + dummy)
    aux = arctan2(z2*d1, p*R12) - arctan2(z1*d1, p*R11)
    res -= g*y1*aux/(p*d1 + dummy) + n*p*aux/(d1 + dummy)
    
    res += n*(log(z2 + R12 + dummy) - log(z1 + R11 + dummy)
           - log(z2 + R22 + dummy) + log(z1 + R21 + dummy))
    res *= -(1.0/(1.0 + n*n + dummy))
    return res
